contrib/parseq-batching/src/main/java/com/linkedin/parseq/batching/BatchingStrategy.java

```diff
@@ -0,0 +1,95 @@
+package com.linkedin.parseq.batching;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+import com.linkedin.parseq.Task;
+import com.linkedin.parseq.internal.ContextImpl;
+import com.linkedin.parseq.internal.PlanContext;
+import com.linkedin.parseq.promise.SettablePromise;
+import com.linkedin.parseq.trace.Relationship;
+import com.linkedin.parseq.trace.ShallowTraceBuilder;
+
+public abstract class BatchingStrategy<K, G> {
+
+  private final ConcurrentHashMap<Long, List<InternalBatchEntry<K, Object>>> batches =
+      new ConcurrentHashMap<>();
+
+  @SuppressWarnings("unchecked")
+  public <T> Task<T> batchable(final String desc, final Callable<BatchEntry<K, T>> callable) {
+    return Task.async("batchable: " + desc, ctx -> {
+      BatchEntry<K, T> batchEntry = callable.call();
+
+      Long planId = ctx.getPlanId();
+      List<InternalBatchEntry<K, Object>> list = batches.get(planId);
+      if (list == null) {
+        list = new ArrayList<>();
+        List<InternalBatchEntry<K, Object>> existing = batches.putIfAbsent(planId, list);
+        if (existing != null) {
+          list = existing;
+        }
+      }
```

Comment:
I think you can avoid two calls to the ConcurrentHashMap here:

```java
Long planId = ctx.getPlanId();
List<InternalBatchEntry<K, Object>> newList = new ArrayList<>();
List<InternalBatchEntry<K, Object>> list = batches.putIfAbsent(planId, newList);
if (list == null) list = newList;
```

Took me a while to figure out that you were doing the second `existing != null` check because of the potential race condition with something getting added to the batch between `.get` and `.putIfAbsent`.